Fix for endpoint not having correct port.

mitchdenny opened the pull request:
Fixes #692

I think this issue was ultimately a very simple bug in the way that we were constructing the allocated endpoint annotation. Whenever the port is specified it should be what is allocated. If it isn't specified then we should take the port allocated.

Note putting this PR up for review ... will remove cruft from sample app host and create a test around this but wanted to get it in front of @karolz-ms especially since this seemed less involved than what he specified in the issue.

mitchdenny committed:
Fix for endpoint not having correct port.

davidfowl commented:
Let’s stop changing the sample and write moar tests

mitchdenny commented:
> Let’s stop changing the sample and write moar tests

They are coming. I always start with experimenting with sample -- its my sandbox :)

davidfowl commented:
We fixed one of the issues running the aspire tests on the ci, but there's another one.

davidfowl commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -167,7 +167,7 @@ private static void AddAllocatedEndpointInfo(IEnumerable<AppResource> resources)
                     PortProtocol.ToProtocolType(svc.Spec.Protocol),
                     svc.AllocatedAddress!,
-                    (int)svc.AllocatedPort!,
+                    (int)(sp.ServiceBindingAnnotation.Port! ?? svc.AllocatedPort!), // Checked by HasCompleteAddress above.
```
```suggestion
                    (int)(sp.ServiceBindingAnnotation.Port ?? svc.AllocatedPort!), // Checked by HasCompleteAddress above.
```

mitchdenny committed:
Update src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

Co-authored-by: David Fowler <davidfowl@gmail.com>

mitchdenny committed:
Add unit tests to cover bug.

mitchdenny commented:
> We fixed one of the issues running the aspire tests on the ci, but there's another one.

What was the other issue? Concurrency of execution with integration tests?

mitchdenny commented:
P.S. Added two covering tests for this issue. Services WithReplicas > 1 would succeed (so I added a test to make sure we don't break that) but anytme you didn't have a replica the service binding annotation and allocated endpoint annotation wouldn't agree on port ... which ultimately was the issue you were seeing. So we have a fix and a covering test to make sure we don't regress that either.

karolz-ms commented:
Looking...

karolz-ms commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -167,7 +167,7 @@ private static void AddAllocatedEndpointInfo(IEnumerable<AppResource> resources)
                     PortProtocol.ToProtocolType(svc.Spec.Protocol),
                     svc.AllocatedAddress!,
-                    (int)svc.AllocatedPort!,
+                    (int)(sp.ServiceBindingAnnotation.Port ?? svc.AllocatedPort!), // Checked by HasCompleteAddress above.
```
This works because if you have just one replica, you can hit the replica directly and it is pretty much the same thing as hitting the proxy for the service that uses a single replica. But this breaks if you start from one replica and then want to scale up to multiple replicas. Granted, we do not have this scenario today in DAB model.

This change also introduces a subtle mismatch between DAB and DCP model: if you use `portFor` environment variable function, it will tell you a different port for a service from what DAB thinks is the port for the same service is (again, for a single-replica service).

But all this said, I think this is fine as an intermediate step.

mitchdenny replied:
The intention behind `AllocatedEndpointAnnotation.Port` is that it is a way for DCP to communicate back to the app model what the front-end port has been allocated to a service. If a service allocates a port via `ServiceBindingAnnotation.Port` then `AllocatedEndpointAnnotation.Port` should always be the same. Where `ServiceBindingAnnotation.Port` is null, that is a signal for DCP to allocate a random port, and that random port should be set in `AllocatedEndpointAnnotation.Port` so that things like environment callbacks can get the actual port to set connection strings etc.

mitchdenny replied:
... I'm not sure how portFor comes into this? I would have thought that at the point that we are constructing the `AllocatedEndpointAnnotation` that we would have this information all setup.

karolz-ms reviewed (commented):
If you agree to roll back the change to eShopLite sample, I am fine with the rest (love the tests!)

The more I think about it, the more I am convinced we should just always use the `ExecutableReplicaSet` for everything and have one "case" for port allocation. So my suggestion would be to merge this and then assign https://github.com/dotnet/aspire/issues/692 to me, and I will make that change. This will position us well for enabling dynamic replica count changes at run time.

mitchdenny commented on samples/eShopLite/AppHost/Program.cs:
```diff
@@ -7,8 +7,7 @@
 var catalogService = builder.AddProject<Projects.CatalogService>("catalogservice")
-                     .WithReference(catalogDb)
-                     .WithReplicas(2);
+                     .WithReference(catalogDb);
```
This was a mistake ... putting it back in :)

mitchdenny committed:
Put replica back.

mitchdenny commented:
@karolz-ms can you approve this?

mitchdenny commented:
@davidfowl  ... @karolz-ms and I caught up on this. My change here is part of the solution, but he'll carrying on the issue with some changes to using replicasets all the time (for consistency).
